components/shared/snapshot/lib.rs

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+}
+
+pub type IpcSnapshot = Snapshot<IpcSharedMemory>;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Snapshot<T = Data> {
+    size: Size2D<u64>,
+    /// internal data (can be any format it will be converted on use if needed)
+    data: T,
+    /// RGBA/BGRA (reflect internal data)
+    format: PixelFormat,
+    /// How to threat alpha channel
+    alpha_mode: AlphaMode,
+}
+
+impl<T> Snapshot<T> {
+    pub const fn size(&self) -> Size2D<u64> {
+        self.size
+    }
+
+    pub const fn format(&self) -> PixelFormat {
+        self.format
+    }
+
+    pub const fn alpha_mode(&self) -> AlphaMode {
+        self.alpha_mode
+    }
+
+    pub const fn is_premultiplied(&self) -> bool {
+        self.alpha_mode().is_premultiplied()
+    }
+
+    pub const fn is_opaque(&self) -> bool {
+        self.alpha_mode().is_opaque()
+    }
+}
+
+impl Snapshot<Data> {
+    pub fn empty() -> Self {
+        Self {
+            size: Size2D::zero(),
+            data: Data::Owned(vec![]),
+            format: PixelFormat::RGBA,
+            alpha_mode: AlphaMode::Transparent {
+                premultiplied: true,
+            },
+        }
+    }
+
+    pub fn cleared(size: Size2D<u64>) -> Self {
+        Self {
+            size,
+            data: Data::Owned(vec![0; size.area() as usize * 4]),
+            format: PixelFormat::RGBA,
+            alpha_mode: AlphaMode::Transparent {
+                premultiplied: true,
+            },
+        }
+    }
+
+    pub fn new(
+        size: Size2D<u64>,
+        format: PixelFormat,
+        alpha_mode: AlphaMode,
+        data: Vec<u8>,
+    ) -> Self {
+        Self {
+            size,
+            data: Data::Owned(data),
+            format,
+            alpha_mode,
+        }
+    }
+
+    pub fn from_vec(
+        size: Size2D<u64>,
+        format: PixelFormat,
+        alpha_mode: AlphaMode,
+        data: Vec<u8>,
+    ) -> Self {
+        Self {
+            size,
+            data: Data::Owned(data),
+            format,
+            alpha_mode,
+        }
+    }
+
+    pub fn from_ism(
+        size: Size2D<u64>,
+        format: PixelFormat,
+        alpha_mode: AlphaMode,
+        ism: IpcSharedMemory,
+    ) -> Self {
+        Self {
+            size,
+            data: Data::Owned(ism.to_vec()),
+            format,
+            alpha_mode,
+        }
+    }
+
+    /*
+    /// # Safety
+    ///
+    /// This is safe is data is owned by this proces only
+    /// (ownership is transferred on send)
+    pub unsafe fn from_ism(
+        size: Size2D<u64>,
+        format: PixelFormat,
+        alpha_mode: AlphaMode,
+        ism: IpcSharedMemory,
+    ) -> Self {
+        Self {
+            size,
+            data: Data::IPC(ism),
+            format,
+            alpha_mode,
+        }
+    }
+    */
+
+    pub fn data(&self) -> &[u8] {
+        &self.data
+    }
+
+    pub fn transform(&mut self, target_alpha_mode: AlphaMode, target_format: PixelFormat) {
```

Comment:
Does the correspond to any spec algorithm? If not, you could still add doc explaining what it is supposed to do?

Reply:
It does conversion to target format / alpha mode. Will write doc.